### DISCUSSION TITLE
[API Request] getCoverArt parameter to return embedded art first

### DISCUSSION THREAD

**Discussion by acroyear** [opensubsonic/open-subsonic-api#42] [API requests]
### Type of change

API Clarification

### Proposal description

In some cases (e.g., Greatest Hits compilations), some of my mp3 files have embedded the single's cover art rather than the 'greatest hits' album. If these songs are all grouped together in the same musicDirectory on the file system, they all get the same cover art. I would like to be able to ask the server to return the embedded coverArt from the song rather than the folder's assigned thumbnail.

### Backward compatibility impact

only an additive feature.

### Backward compatibility

- [X] No backward compatibility impact.

### API details

```/rest/getCoverArt?size=<size>&id=<id>&preferEmbedded=<true|false>```

### Security impacts

Can't be sure if there's not some XSS corrupting info in the image, of course, but that's true even for what the java code is extracting.

### Potential issues

If the data is cached with the rest of the thumbnails, it can greatly increase the size of the thumbnail folder on the server. Maybe enabling the feature could be a configuration option.

### Alternative solutions

Alternate solution is for the server itself to have configurations to decide if a song's thumbnail should return the folder or the song's individual rather than an API call. Not possible with subsonic or its derivatives currently, of course. But the API version could give the individual user the to (for clients supporting it) the ability to switch between them.

**Comment by dweymouth** [comment]
I think this is unnecessary and it's an implementation detail for the server. Not all servers support embedded art. Of the ones that do, some (eg Navidrome) let the cover art priority be configured as a server setting.

**Reply by acroyear** [reply]
missing my point. "not all server" - fine. but for those that do, i would prefer a preference settings.

the current attitude is that it is out of the client's control what they see - a server could return embedded, or a referenced locally host 'cover' file, or...anything. there is no way for a client to set the preferences.

i want the server to have the ability to provide alternatives, and a client the ability to set what they would prefer, between what the folder has vs what the mp3/flac file actually contains.

re-opening.

**Reply by acroyear** [reply]
THE WHOLE POINT OF THIS PAGE is to discuss the API and how to account for desired features while accommodating (backwards compatibility) the existing server.

rejecting such a request wholesale like this without others adding to the discussion is the antithesis of that aim, in my opinion.

how do we make the API better for new clients and new servers? this request is very much a part of that.

**Reply by dweymouth** [reply]
Alright I'll change the status back to in discussion, but I don't think you'll get much support for this idea from the other server and client devs. In general, the server's job is to figure out *where* the metadata comes from and the client just requests it based on *what* it needs. eg. servers can get artist images from LastFM or Spotify, artist bios can come from local .nfo files or from LastFM or other sources - all that's hidden from the clients by design and usually the servers have server configuration options to set preferred image or metadata sources. For example, Navidrome has a configuration variable to change the cover art priority order among embedded, folder.[jpg|png], cover.[jpg|png], \<albumname\>.[jpg|png], external, etc

**Reply by Tolriq** [reply]
getCovertArt should not be changed, as the fixed documentation explains it's purpose is to return an image from an imageID.
A single imageID represent a single image it can't represent multiple images else caching on client side is impossible since we don't know if the parameter is applied or not and what is returned.

There's another extension opened to support returning multiple different images that's what is needed and cover your need too.
